nilearn/regions/region_extractor.py

```diff
@@ -0,0 +1,487 @@
+        then mask image to that particular brain region should be provided.
+        If none, masking will not be applied.
+
+    Returns
+    -------
+    regions_extracted: a 4D Nifti-like image
+        contains the images of segmented regions each 3D image is a
+        seperate brain activated region or atlas region.
+
+    indentity_id_of_maps: a numpy array
+        an array of list of indices where each index value is assigned to
+        each seperate region of its corresponding family of brain maps.
+    """
+    all_regions_imgs = []
+    identity_id_of_maps = []
+    maps_img = check_niimg(maps_img, atleast_4d=True)
+    maps = maps_img.get_data()
+    affine = maps_img.get_affine()
+
+    extract_methods = ['connected_components', 'local_regions']
+    if extract_type not in extract_methods:
+        message = ("'extract_type' should be given "
+                   "either of these {0}").format(extract_methods)
+        raise ValueError(message)
+
+    if mask_img is not None:
+        if not _check_same_fov(maps_img, mask_img):
+            mask_img = resample_img(mask_img,
+                                    target_affine=maps_img.get_affine(),
+                                    target_shape=maps_img.shape[:3],
+                                    interpolation="nearest")
+            mask = masking._load_mask_img(mask_img)
+            # Set as 0 to the values which are outside of the mask
+            maps[mask == 0.] = 0.
+
+    for index in range(maps.shape[-1]):
+        regions = []
+        map_ = maps[..., index]
+        # Mark the seeds using random walker
+        if extract_type == 'local_regions':
+            smooth_map = _smooth_array(map_, affine, peak_local_smooth)
+            seeds = peak_local_max(smooth_map, indices=False,
+                                   exclude_border=False)
+            seeds_label, seeds_id = label(seeds)
+            # Assign -1 to values which are 0. to indicate to ignore
+            seeds_label[map_ == 0.] = -1
+            # Take the maximum value of the data and
+            # normalize to be between -1 and 1 for random walker compatibility
+            max_value_map = max(map_.max(), -map_.min())
+            if max_value_map > 1.:
+                map_ /= max_value_map
+
+            try:
+                mode = 'cg_mg'
+                rw_maps = random_walker(map_, seeds_label, mode=mode)
+            except Exception as e:
+                print("Random Walker algorithm failed for mode:%s, %s" % (mode, str(e)))
+                rw_maps = random_walker(map_, seeds_label, mode='bf')
+
+            # Now simply replace "-1" with "0" for regions seperation
+            rw_maps[rw_maps == -1] = 0.
+            maps_assigned = rw_maps
+        else:
+            # Taking directly the input data which is the case in 'auto' type
+            maps_assigned = map_
+
+        # Region seperation block
+        label_maps, n_labels = label(maps_assigned)
```

Comment:
I think you don't need that if you have run RW before. Therefore it should go into the `else` statement.